src/fswAlgorithms/attGuidance/sunSafePointCpp/sunSafePointCpp.cpp

```diff
@@ -62,23 +61,71 @@ void SunSafePointCpp::Reset(uint64_t callTime)
     if (this->sunDirectionInMsg.isWritten()) {
-        sunDirectionInBuffer = this->sunDirectionInMsg();
+        this->sunDirectionInBuffer = this->sunDirectionInMsg();
+    }
+
+    // Determine norm of measured Sun-direction vector
+    double sHatNorm = cArray2EigenVector3d(this->sunDirectionInBuffer.vehSunPntBdy).norm();
+
+    // Computing the attitude guidance states sigma_BR and omega_RN_B
+    if (this->sunDirectionIsAvailable(sHatNorm)) {
+        this->computeAttGuidanceStates(sHatNorm);
+    } else {
+        v3SetZero(this->attGuidanceOutBuffer.sigma_BR);
     }
 
+    // Compute the hub angular rate error omega_BR_B
+    this->computeHubAngularRateError();
+
+    // Write the guidance output message
+    eigenVector3d2CArray(this->omega_RN_B, this->attGuidanceOutBuffer.omega_RN_B);
+    this->attGuidanceOutMsg.write(&this->attGuidanceOutBuffer, moduleID, callTime);
+}
+
+/*! Method for computing the attitude guidance states sigma_BR and omega_RN_B if a valid sun direction vector is available.
+ @return void
+ @param sHatNorm Norm of measured Sun-direction vector
+*/
+void SunSafePointCpp::computeAttGuidanceStates(double sHatNorm) {
+    double dotProductNormalized = (this->sHatBdyCmd.dot(cArray2EigenVector3d(this->sunDirectionInBuffer.vehSunPntBdy)))
+                                          / sHatNorm;
+    dotProductNormalized = fabs(dotProductNormalized) > 1.0 ?
+    dotProductNormalized/fabs(dotProductNormalized) : dotProductNormalized;
+    this->sunAngleErr = safeAcos(dotProductNormalized);
+
+    // Compute the heading error relative to the sun direction vector
+    Eigen::Vector3d e_hat;  // Eigen Axis
+    if (this->sunAngleErr < this->smallAngle) {  // Sun heading and desired body axis are essentially aligned. Set attitude error to zero.
+        v3SetZero(this->attGuidanceOutBuffer.sigma_BR);
+    } else {
+        if (M_PI - this->sunAngleErr < this->smallAngle) {  // The commanded body vector nearly is opposite the sun heading
+            e_hat = this->eHat180_B;
+        } else {  // Normal case where sun and commanded body vectors are not aligned
+            e_hat = cArray2EigenVector3d(this->sunDirectionInBuffer.vehSunPntBdy).cross(this->sHatBdyCmd);
+        }
+        this->sunMnvrVec = e_hat / e_hat.norm();
+        Eigen::Vector3d v2 = tan(this->sunAngleErr * 0.25) * this->sunMnvrVec;
+        eigenVector3d2CArray(v2, this->attGuidanceOutBuffer.sigma_BR);
+        MRPswitch(this->attGuidanceOutBuffer.sigma_BR, 1.0, this->attGuidanceOutBuffer.sigma_BR);
+    }
+
+    // Rate tracking error is the body rate to bring spacecraft to rest
+    this->omega_RN_B = (this->sunAxisSpinRate / sHatNorm) * cArray2EigenVector3d(this->sunDirectionInBuffer.vehSunPntBdy);
+}
+
+/*! Method for computing the hub angular rate error omega_BR_B.
+ @return void
+*/
+void SunSafePointCpp::computeHubAngularRateError() {
     // Read the imu guidance input message
+    NavAttMsgPayload localImuDataInBuffer;
```

Comment:
:large_blue_circle: We can declare and initialize message variables on a single line. E.g. `NavAttMsgPayload localImuDataInBuffer =  NavAttMsgPayload();`

Reply:
Done